feat(ecosystem): Adds Project Management integration SLOs

GabeVillalobos opened the pull request:
# Adds Project Management/Issues SLO metrics

This PR adds metric coverage to the majority of Project Management features:
- Issue Creation (including alert rules) - src/sentry/api/endpoints/group_integration_details.py
- Outbound status syncing - src/sentry/integrations/tasks/sync_status_outbound.py
- Outbound assignment syncing - src/sentry/integrations/tasks/sync_assignee_outbound.py
- Issue linking - src/sentry/api/endpoints/group_integration_details.py
- Inbound assignment syncing
- Installation metrics for ticketing integrations
- 
## Has Metrics, Needs Testing:
- Inbound status syncing


## Incidental Work:
- Adds new tests for `sync_outbound_assignment` and `sync_outbound_status` tasks, which previously only had endpoint level testing.
- Adds new test for the shared `sync_group_assignee_inbound` utility (src/sentry/integrations/utils/sync.py)

GabeVillalobos committed:
feat(ecosystem): Adds issue SLOs along with issue provider names

GabeVillalobos committed:
Updates metrics interface to take integration, not installation

GabeVillalobos committed:
Adds new tests plus metrics assertions for ticketing

GabeVillalobos committed:
Adds new metrics assertion util file

GabeVillalobos committed:
Removes previously added issue provider name

GabeVillalobos committed:
Adds sync_status_outbound test

GabeVillalobos committed:
Removes unnecessary util, updates tests to assert on metrics method invocations

GabeVillalobos committed:
Adds tests for outbound assignee sync task

GabeVillalobos commented on src/sentry/integrations/project_management/metrics.py:
```diff
@@ -0,0 +1,35 @@
+@dataclass
+class ProjectManagementEvent(IntegrationEventLifecycleMetric):
+    action_type: ProjectManagementActionType
+    integration: Integration | RpcIntegration
```
Turns out we don't always have access to integration installations (specifically the inbound assignment code), so I made this use integration/rpc integration.

GabeVillalobos commented on src/sentry/rules/actions/integrations/create_ticket/utils.py:
```diff
@@ -114,27 +118,38 @@ def create_issue(event: GroupEvent, futures: Sequence[RuleFuture]) -> None:
+            try:
+                response = installation.create_issue(data)
+            except Exception as e:
+                logger.info(
+                    "%s.rule_trigger.create_ticket.failure",
+                    provider,
+                    extra={
+                        "rule_id": rule_id,
+                        "provider": provider,
+                        "integration_id": integration.id,
+                        "error_message": str(e),
+                        "exception_type": type(e).__name__,
+                    },
```
It's nice being able to delete these logs in the future

GabeVillalobos commented on tests/sentry/integrations/jira/test_ticket_action.py:
```diff
@@ -1,29 +1,48 @@
 pytestmark = [requires_snuba]
 
 
+class BrokenJiraMock(MockJira):
+    """
+    Subclass of MockJira client, which implements a broken create_issue method
+    to simulate a misconfigured alert rule for Jira.
+    """
+
+    def create_issue(self, raw_form_data):
+        raise ApiInvalidRequestError("Invalid data entered")
```
This is gross, but was the easiest way to trigger a failure scenario.

iamrajjoshi commented on src/sentry/api/endpoints/group_integration_details.py:
```diff
@@ -168,56 +172,66 @@ def put(self, request: Request, group, integration_id) -> Response:
+            try:
+                installation.after_link_issue(external_issue, data=request.data)
+            except IntegrationFormError as exc:
+                lifecycle.record_halt(exc)
```
are these errors user config errors?

GabeVillalobos replied:
Yeah, typically an `IntegrationFormError` is an incorrectly configured Jira ticket creation config.

iamrajjoshi commented on src/sentry/integrations/jira/utils/api.py:
```diff
@@ -52,27 +53,34 @@ def handle_assignee_change(
+            item for item in data["changelog"]["items"] if item["field"] == "assignee"
+        )
+        if not assignee_changed:
+            lifecycle.record_halt()
```
could this possibly be a success? are these normal scenarios where jira doesn't send us assignee?

GabeVillalobos replied:
That's fair. Tbh though, I'm looking at moving this call into the shared util this calls as there are specific lifecycles we probably want to track there, so this is likely to change.

iamrajjoshi commented on src/sentry/integrations/jira/utils/api.py:
```diff
@@ -52,27 +53,34 @@ def handle_assignee_change(
+        assignee = fields.get("assignee")
+
+        if assignee is None:
+            lifecycle.record_halt()
```
does this mean the ticket was unassigned? why does this have to be a halt?

GabeVillalobos replied:
Yep, this is actually fine and should be recorded as a success.

iamrajjoshi reviewed (commented):
i just have a few questions, but i think we are kind of aligned on "halt".

i have been thinking of fault as any unexpected condition that isn't user's fault.
with that definition, i have been considering halt as all other non-success conditions, which makes it probably the broadest in definition.

GabeVillalobos committed:
Updates record_halts to include messages, fixes broken test

Also resets changes made to jira utils API file, which is probably the
incorrect place to be adding metrics gathering.

GabeVillalobos committed:
Adds tests for sync_assignee_inbound util

GabeVillalobos committed:
Fixes typing issues, adds assertion for metric gathering in group integration details

GabeVillalobos commented on src/sentry/api/endpoints/group_integration_details.py:
```diff
@@ -168,56 +172,66 @@
+                return Response(exc.field_errors, status=400)
+            except IntegrationError as e:
+                lifecycle.record_halt(e)
+                return Response({"non_field_errors": [str(e)]}, status=400)
```
This isn't new code, this is what allows errors to propagate up to the UI when ticket creation or linking fails on the integrator side (usually a config issue with the ticket contents)

GabeVillalobos committed:
Adds metric gathering to inbound status sync flows

GabeVillalobos committed:
Removes accidental whitespace

GabeVillalobos committed:
Adds Jira installation metrics

GabeVillalobos committed:
Adds tests for linking metrics, untested branches

GabeVillalobos committed:
Fixes bug with project manager action context

iamrajjoshi commented on src/sentry/integrations/jira/utils/api.py:
```diff
@@ -75,26 +76,40 @@ def handle_assignee_change(
+                item for item in data["changelog"]["items"] if item["field"] == "status"
+            )
+        except StopIteration:
+            lifecycle.record_halt("missing-changelog-status", extra=log_context)
```
extreme nit - do we want to put all the reasons in an enum or something?

GabeVillalobos replied:
Fair! I'll do that now.

iamrajjoshi commented on src/sentry/integrations/utils/sync.py:
```diff
@@ -77,48 +81,52 @@ def sync_group_assignee_inbound(
+                groups_assigned.append(group)
+            else:
+                lifecycle.record_halt("inbound-assignee-not-found", extra=log_context)
+                logger.info("inbound-assignee-not-found", extra=log_context)
```
we should save a cleanup task for ourselves to remove these redundant loggers

GabeVillalobos replied:
Good call. Adding it to our SLOs tasks

iamrajjoshi commented on src/sentry/rules/actions/integrations/create_ticket/utils.py:
```diff
@@ -114,27 +118,36 @@ def create_issue(event: GroupEvent, futures: Sequence[RuleFuture]) -> None:
+            action_type=ProjectManagementActionType.CREATE_EXTERNAL_ISSUE,
+            integration=integration,
+        ).capture() as lifecycle:
+            lifecycle.add_extra("provider", provider)
```
i was planning to add a quick little utility to remove these repeat calls  in my pr here:
https://github.com/getsentry/sentry/pull/80984/files?w=0#r1851251413.

i can update this after

iamrajjoshi reviewed (approved):
lgtm

GabeVillalobos commented on src/sentry/api/endpoints/group_integration_details.py:
```diff
@@ -168,56 +172,66 @@
+                return Response(exc.field_errors, status=400)
+            except IntegrationError as e:
+                lifecycle.record_halt(e)
+                return Response({"non_field_errors": [str(e)]}, status=400)
```
Same thing here, existing code.

GabeVillalobos committed:
Adds enum to consolidate halt reasons

GabeVillalobos committed:
Fixes test with broken assertion
